src/layout/umap.c

```diff
@@ -0,0 +1,1047 @@
+        xm += MATRIX(*layout, i, 0);
+        ym += MATRIX(*layout, i, 1);
+    }
+    xm /= no_of_nodes;
+    ym /= no_of_nodes;
+
+    /* Shift vertices */
+    for (igraph_integer_t i = 0; i < no_of_nodes; i++) {
+        MATRIX(*layout, i, 0) -= xm;
+        MATRIX(*layout, i, 1) -= ym;
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* Check the distances argument, which should be NULL or a vector of nonnegative numbers */
+static igraph_error_t igraph_i_umap_check_distances(const igraph_vector_t *distances, igraph_integer_t no_of_edges) {
+
+    if (distances == NULL) {
+        return IGRAPH_SUCCESS;
+    }
+
+    if (igraph_vector_size(distances) != no_of_edges) {
+        IGRAPH_ERROR("Distances must be the same number as the edges in the graph.", IGRAPH_EINVAL);
+    }
+
+    for (igraph_integer_t eid = 0; eid != no_of_edges; eid++) {
+        if (VECTOR(*distances)[eid] < 0) {
+            IGRAPH_ERROR("Distances cannot be negative.", IGRAPH_EINVAL);
+        } else if (igraph_is_nan(VECTOR(*distances)[eid])) {
+            IGRAPH_ERROR("Distances cannot contain NaN values.", IGRAPH_EINVAL);
+        }
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/* This is the main function that works for any dimensionality of the embedding (currently hard-constrained to 2 or 3 ONLY in the initialization) */
+static igraph_error_t igraph_i_layout_umap(const igraph_t *graph, const igraph_vector_t *distances,
+        igraph_matrix_t *layout, igraph_real_t min_dist, igraph_integer_t epochs, igraph_real_t sampling_prob, igraph_integer_t ndim) {
+
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    /* probabilities of each edge being a real connection */
+    igraph_vector_t umap_weights;
+    /* The smoothing parameters given min_dist */
+    igraph_real_t a, b;
+
+    /* Check input arguments */
+    if (min_dist <= 0) {
+        IGRAPH_ERRORF("Minimum distance should be positive, but found %g.",
+                IGRAPH_EINVAL, min_dist);
+    }
+
+    if (epochs < 0) {
+        IGRAPH_ERRORF("Number of epochs should be non-negative, but found %"
+                IGRAPH_PRId ".",
+                IGRAPH_EINVAL, epochs);
+    }
+
+    if (sampling_prob <= 0) {
+        IGRAPH_ERRORF("Sampling probability should be positive, but found %g.",
+                IGRAPH_EINVAL, sampling_prob);
+    }
+
+    if ((ndim != 2) && (ndim != 3)) {
+        IGRAPH_ERRORF("Number of dimensions should be 2 or 3, found %d.",
+                IGRAPH_EINVAL, (int)ndim);
+
+    }
+
+    /* UMAP is sometimes used on unweighted graphs, that means distances are always zero */
+    IGRAPH_CHECK(igraph_i_umap_check_distances(distances, no_of_edges));
+
+    /* Trivial graphs (0 or 1 nodes) beget trivial - but valid - layouts */
+    if (no_of_nodes <= 1) {
+        IGRAPH_CHECK(igraph_matrix_resize(layout, no_of_nodes, ndim));
+        if (no_of_nodes == 1) {
```

Comment:
@GroteGnoom @iosonofabio Why did you use this conditional here? Does `matrix_null` not work for zero-size matrices? If so, that would be a bug that needs to be fixed.

Reply:
As far as I can tell, there's no problem with calling `matrix_null` on zero-size matrices. I'll remove the conditional so we wouldn't keep wondering about it next time we see it. Let me know if I missed something.